File: snippets/python/algo/general/algo_merge_sort.py
def merge_sort(seq):
	if len(seq)<2: return seq
	n = len(seq)//2
	left, right = merge_sort(seq[:n]), merge_sort(seq[n:])
	i, j, n, m, seq = 0, 0, len(left), len(right), []

	while i < n and j < m:
		if left[i] < right[j]: # <Comparison>
			seq.append(left[i])
			i += 1
		else:
			seq.append(right[j])
			j += 1
	seq.extend(left[i:])
	seq.extend(right[j:])
	return seq

$0

File: snippets/cpp/algo/general/algo_merge_sort.cpp
//@title Merge Sort algorithme
//@defines mergerSort
//@doc Can be adapted for any purpose (for exemple, any algorithm based on divide-to-conquer)
//@
template<class T> T mergeSort(T elements) {
	if (elements.size() >= 2) {
		T left, right;
		for (int iElement = 0; iElement < (int)elements.size()/2; iElement++) {
			left.push_back(elements[iElement]);
		}
		for (int iElement = (int)elements.size()/2; iElement < (int)elements.size(); iElement++) {
			right.push_back(elements[iElement]);
		}
		left = mergeSort(left);
		right = mergeSort(right);

		int i1 = 0, i2 = 0;
		while (i1 < (int)left.size()) {
			if (i2 >= (int)right.size() || left[i1] < right[i2]) {
				elements[i1+i2] = left[i1];
				i1++;
			} else {
				elements[i1+i2] = right[i2];
				i2++;
			}
		}
		while (i2< (int)right.size()) {
			elements[i1+i2] = right[i2];
			i2++;
		}
	}
	return elements;
}

